Objects/codeobject.c

```diff
@@ -333,55 +568,224 @@ PyCode_New(int argcount, int kwonlyargcount,
+    };
+    result = _PyCode_New(&con);
 
 failed:
+    Py_XDECREF(emptystring);
+    Py_XDECREF(nulltuple);
     Py_XDECREF(funcname_ob);
     Py_XDECREF(filename_ob);
     return result;
 }
 
 
+/******************
+ * fast locals
+ ******************/
+
+bool
+_PyCode_HasFastlocals(PyCodeObject *co, _PyFastLocalKind kind)
+{
+    if (kind & CO_FAST_POSONLY && co->co_posonlyargcount > 0) {
+        return true;
+    }
+    if (kind & CO_FAST_POSORKW &&
+            co->co_argcount - co->co_posonlyargcount > 0) {
+        return true;
+    }
+    if (kind & CO_FAST_VARARGS && co->co_flags & CO_VARARGS) {
+        return true;
+    }
+    if (kind & CO_FAST_KWONLY && co->co_kwonlyargcount > 0) {
+        return true;
+    }
+    if (kind & CO_FAST_VARKWARGS && co->co_flags & CO_VARKEYWORDS) {
+        return true;
+    }
+
+    if (kind & CO_FAST_LOCALONLY) {
+        int totalargs = co->co_argcount +
+                        co->co_kwonlyargcount +
+                        ((co->co_flags & CO_VARARGS) != 0) +
+                        ((co->co_flags & CO_VARKEYWORDS) != 0);
+        if (co->co_nlocals - totalargs) {
+            return true;
+        }
+    }
+
+    if (kind & CO_FAST_CELL && co->co_ncellvars) {
+        return true;
+    }
+
+    if (kind & CO_FAST_FREE && co->co_nfreevars) {
+        return true;
+    }
+
+    return false;
+}
+
+void
+_PyCode_FastInfoFromOffset(PyCodeObject *co, int offset,
+                           PyObject **pname, _PyFastLocalKind *pkind)
+{
+    assert(offset >= 0);
+    assert(pname != NULL || pkind != NULL);
+
+    PyObject *names;
+    int index = offset;
+    _PyFastLocalKind kind;
+    if (index < co->co_nlocals) {
+        names = co->co_varnames;
+        // For now we do not distinguish arg kinds.
+        kind = CO_FAST_LOCAL;
+    }
+    else {
+        index -= co->co_nlocals;
+        if (index < co->co_ncellvars) {
+            names = co->co_cellvars;
+            kind = CO_FAST_CELL;
+        }
+        else {
+            index -= co->co_ncellvars;
+            assert(index < co->co_nfreevars);
+            names = co->co_freevars;
+            kind = CO_FAST_FREE;
+        }
+    }
+
+    if (pname != NULL) {
+        *pname = PyTuple_GET_ITEM(names, index);
+    }
+    if (pkind != NULL) {
+        *pkind = kind;
+    }
+}
+
+void
+_PyCode_FastInfoFromOparg(PyCodeObject *co, int oparg,
+                          _PyFastLocalKind expected,
+                          PyObject **pname, _PyFastLocalKind *pkind)
+{
+    assert(expected > 0);
+    if (expected & CO_FAST_LOCAL) {
+        assert(expected < CO_FAST_CELL);
+        assert(oparg < co->co_nlocals);
+        _PyCode_FastInfoFromOffset(co, oparg, pname, pkind);
+    }
+    else {
+        // For now, opargs for cells and free vars are separate from locals.
+        _PyCode_FastInfoFromOffset(co, oparg + co->co_nlocals, pname, pkind);
+    }
+}
+
+int
+_PyCode_OffsetFromIndex(PyCodeObject *co, int index, _PyFastLocalKind kind)
```

Comment:
I don't like the use of this function everywhere.
If inlined, it should be no slower than the code it is replacing, but what is the point?
`index + co->co_nlocals` is clearer to me than `_PyCode_OffsetFromIndex(co, index, CO_FAST_CELL)`.
`index` is far, far clearer than ` `_PyCode_OffsetFromIndex(co, index, CO_FAST_LOCAL)`